File: client/app/src/main/java/com/jackpf/locationhistory/client/ssl/SSLUtils.java
package com.jackpf.locationhistory.client.ssl;

import java.security.MessageDigest;
import java.security.cert.X509Certificate;

public class SSLUtils {
    public static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < bytes.length; i++) {
            if (i > 0) sb.append(":");
            sb.append(String.format("%02X", bytes[i]));
        }
        return sb.toString();
    }

    public static String getSha256Fingerprint(X509Certificate cert) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] publicKey = md.digest(cert.getEncoded());
            return bytesToHex(publicKey);
        } catch (Exception e) {
            throw new RuntimeException("Failed to generate SHA-256 fingerprint", e);
        }
    }
}

File: server/src/main/scala/com/jackpf/locationhistory/server/util/SSLUtils.scala
package com.jackpf.locationhistory.server.util

import java.io.{File, FileInputStream}
import java.security.MessageDigest
import java.security.cert.{CertificateFactory, X509Certificate}
import scala.util.{Failure, Success, Using}

object SSLUtils {
  def sslFingerprint(certFile: File): String = {
    Using(new FileInputStream(certFile)) { inputStream =>
      val certificateFactory = CertificateFactory.getInstance("X.509")
      val cert = certificateFactory.generateCertificate(inputStream).asInstanceOf[X509Certificate]

      val digest = MessageDigest.getInstance("SHA-256")
      val hash = digest.digest(cert.getEncoded)
      val hex = hash.map("%02X".format(_)).mkString(":")

      s"""${"=" * 50}
         |SERVER FINGERPRINT (SHA-256):
         |${hex}
         |${"=" * 50}""".stripMargin
    }
  } match {
    case Failure(exception) => throw exception
    case Success(v)         => v
  }
}
